### PR TITLE
Forum Edit Categories (Reordering)

### PR DESCRIPTION
**Pull request by scopeInfinity**
Issue: #2094 (Forum: color coded & instructor ordered tags)

To be merged after PR Forum allow multiple categories for a thread  #2083, need to change Base Branch of PR before merging.

- Categories Reordering
  - Used Sortable JS for UI
  - After successful reordering, refresh categories order on same page without making additional request.

### CODE REVIEW

File: .setup/update_database.py [hunk 1]
```diff
@@ -122,6 +122,9 @@
         os.system("PGPASSWORD='{}' psql --host={} --username={} --dbname={} -c 'ALTER TABLE ONLY threads ADD COLUMN merged_thread_id int DEFAULT '-1''".format(*variables))
         os.system("PGPASSWORD='{}' psql --host={} --username={} --dbname={} -c 'ALTER TABLE ONLY threads ADD COLUMN merged_post_id int DEFAULT '-1''".format(*variables))
         
+        # categories ordering for discussion forum
+        os.system("PGPASSWORD='{}' psql --host={} --username={} --dbname={} -c 'ALTER TABLE ONLY categories_list ADD COLUMN rank int'".format(*variables))
+
```
Comment (andrewaikens87) [inline .setup/update_database.py]: Any database change also has to be added into `course_tables.sql`.

File: site/app/views/forum/ForumThreadView.php [hunk 2]
```diff
@@ -847,7 +848,54 @@ public function createThread() {
...
+			$categories = $this->core->getQueries()->getCategories();
+			$return .= <<<HTML
+			<div class="popup-form" id="category-list">
+				<h3>Categories</h3>
+				<pre>(Drag to re-order)</pre><br>
+HTML;
+				if(count($categories) == 0) {
+					$return .= <<<HTML
+					<span id='category-list-no-element' style="margin-left: 1em;" >
+						No category found
```
Comment (andrewaikens87) [inline site/app/views/forum/ForumThreadView.php]: If there is no category found there still displays a draggable bar. Can we please remove that? Also if there exists no categories we should instruct the user to create one. So possibly something like "No categories exist please create one." or something similar. 


draggablebar

File: site/public/js/server.js [hunk 3]
```diff
@@ -1270,6 +1270,101 @@ function addNewCategory(){
...
+
+function reorderCategories(){
+    var data = $('#ui-category-list').sortable('serialize');
+    var url = buildUrl({'component': 'forum', 'page': 'reorder_categories'});
+    $.ajax({
+            url: url,
+            type: "POST",
+            data: data,
+            success: function(data){
+                console.log(data);
```
Comment (andrewaikens87) [inline site/public/js/server.js]: Please remove console log

**Comment by bmcutler** [comment]
There should always be at least one category.  We start with "question" & "comment" (they can be renamed/removed if desired, as long as there is always at least one category).

Only instructor/fullaccess/limitedaccess can create or delete categories.

Only unused categories can be deleted.

You cannot remove the last category.  (a category is required to make a post, and a student cannot make a category)

**Comment by andrewaikens87** [comment]
Yes, there should always be at least one category. The error message should only show up if someone manually deleted the rows in the `categories_list` table. This PR currently has changes for altering the ordering of the categories. Renaming/deleting categories hasn't been implemented yet.

File: site/app/views/forum/ForumThreadView.php (hunk 2 above)
Comment (scopeInfinity) [reply site/app/views/forum/ForumThreadView.php]: Thanks! Bug fixed.

**Pull request merged**